Renaming layers from inspector

jmb462 opened the pull request:
###  Proposal implemented

Fixes #57364 (using ``project_settings_changed`` signal)
This closes https://github.com/godotengine/godot-proposals/issues/3275

Right click on a layer brings a Popmenu to rename this layer.
Click this menu opens a ConfirmDialog with the current name, allowing to change it (layer index is shown in the title)
When confirming the rename, tooltips, name list in inspector and project properties are updated with the new layer name.

![rename_layers](https://user-images.githubusercontent.com/3649998/132921900-0095fb2d-c022-45ee-a8b4-5d6602afa19b.gif)

### About the file changes

As it grows with my modification, I moved the EditorPropertyLayersGrid class declaration in editor_properties.h (the class was defined and implemented in the cpp file). That's why Github seems to find more modification that I've really done.

The noticeable changes are : 

- adding two callback function in EditorPropertyLayersGrid : _rename_pressed() and _rename_operation_confirm()
https://github.com/godotengine/godot/blob/100bd36237b92929ffd562f5bc691967d80fd463/editor/editor_properties.cpp#L733-L757
- handling the right click event in EditorPropertyLayersGrid::gui_input() to trigger the popup.
https://github.com/godotengine/godot/blob/100bd36237b92929ffd562f5bc691967d80fd463/editor/editor_properties.cpp#L857-L864
- add a method to rename a layer : EditorPropertyLayers::set_layer_name
https://github.com/godotengine/godot/blob/100bd36237b92929ffd562f5bc691967d80fd463/editor/editor_properties.cpp#L1095-L1100
- using a signal to communicate from EditorPropertyLayersGrid to its grand-parent EditorPropertyLayers

pycbouh commented:
Please use a better commit message than that 🙃

akien-mga commented on editor/editor_properties.cpp:
```diff
@@ -730,265 +730,294 @@ EditorPropertyFlags::EditorPropertyFlags() {
+	rename_dialog_text->set_text(name);
+	rename_dialog_text->select(0, name.length());
+	rename_dialog->popup_centered(Size2(300, 80) * EDSCALE);
+	rename_dialog_text->grab_focus();
+}
+void EditorPropertyLayersGrid::_rename_operation_confirm() {
+	String new_name = rename_dialog_text->get_text().strip_edges();
+	if (new_name.length() == 0) {
+		EditorNode::get_singleton()->show_warning(TTR("No name provided."));
+		return;
+	} else if (new_name.find("/") != -1 || new_name.find("\\") != -1 || new_name.find(":") != -1) {
+		EditorNode::get_singleton()->show_warning(TTR("Name contains invalid characters."));
+		return;
+	}
+	names.set(renamed_layer_index, new_name);
+	tooltips.set(renamed_layer_index, new_name + "\n" + vformat(TTR("Bit %d, value %d"), renamed_layer_index, 1 << renamed_layer_index));
+	emit_signal(SNAME("rename_confirmed"), renamed_layer_index, new_name);
+}
+Size2 EditorPropertyLayersGrid::get_grid_size() const {
+	Ref<Font> font = get_theme_font(SNAME("font"), SNAME("Label"));
+	int font_size = get_theme_font_size(SNAME("font_size"), SNAME("Label"));
+	return Vector2(0, font->get_height(font_size) * 3);
+}
+EditorPropertyLayersGrid::EditorPropertyLayersGrid() {
```
```suggestion
	rename_dialog_text->grab_focus();
}

void EditorPropertyLayersGrid::_rename_operation_confirm() {
	String new_name = rename_dialog_text->get_text().strip_edges();
	if (new_name.length() == 0) {
		EditorNode::get_singleton()->show_warning(TTR("No name provided."));
		return;
	} else if (new_name.find("/") != -1 || new_name.find("\\") != -1 || new_name.find(":") != -1) {
		EditorNode::get_singleton()->show_warning(TTR("Name contains invalid characters."));
		return;
	}
	names.set(renamed_layer_index, new_name);
	tooltips.set(renamed_layer_index, new_name + "\n" + vformat(TTR("Bit %d, value %d"), renamed_layer_index, 1 << renamed_layer_index));
	emit_signal(SNAME("rename_confirmed"), renamed_layer_index, new_name);
}

Size2 EditorPropertyLayersGrid::get_grid_size() const {
	Ref<Font> font = get_theme_font(SNAME("font"), SNAME("Label"));
	int font_size = get_theme_font_size(SNAME("font_size"), SNAME("Label"));
	return Vector2(0, font->get_height(font_size) * 3);
}

EditorPropertyLayersGrid::EditorPropertyLayersGrid() {
```

akien-mga replied:
Not very visible on the diff - I'm adding a line of separation between all method implementations.

jmb462 replied:
Thanks. Done.

jmb462 committed:
Renaming layers from the inspector via a popup menu.

jmb462 commented:
As the conflicted changes with master since 2 month (last rebase) was quite awful to manage, I completely rewrite the changes from the current master.

As it grows with my modification, I moved the EditorPropertyLayersGrid class declaration in editor_properties.h (the class was defined and implemented in the cpp file). That's why git seems to find more modification that I've really done.

@akien-mga Maybe, i should keep declaration of EditorPropertyLayersGrid in the cpp file so my real changes will be easier to review.
Then, we could move class declaration in .h in a second time (or not).

akien-mga commented on editor/editor_properties.cpp:
```diff
@@ -790,262 +790,293 @@ EditorPropertyFlags::EditorPropertyFlags() {
+		return;
+	}
+	String name = names[renamed_layer_index];
+	rename_dialog->set_title(TTR("Renaming layer " + itos(renamed_layer_index + 1) + ":"));
```
Partial strings can't be translated properly, you need to use string formatting with placeholders for this:
```suggestion
	rename_dialog->set_title(vformat(TTR("Renaming layer %d:"), renamed_layer_index + 1));
```

akien-mga commented on editor/editor_properties.cpp:
```diff
@@ -790,262 +790,293 @@ EditorPropertyFlags::EditorPropertyFlags() {
 }
 
 ///////////////////// LAYERS /////////////////////////
+void EditorPropertyLayersGrid::_rename_pressed(int p_menu) {
```
Clearer with keeping the separation:
```suggestion
///////////////////// LAYERS /////////////////////////

void EditorPropertyLayersGrid::_rename_pressed(int p_menu) {
```

akien-mga commented on editor/editor_properties.cpp:
```diff
@@ -790,262 +790,293 @@ EditorPropertyFlags::EditorPropertyFlags() {
+		if (hovered_index >= 0) {
+			renamed_layer_index = hovered_index;
+			layer_rename->set_position(get_screen_position() + mb->get_position());
+			layer_rename->set_size(Vector2(1, 1));
```
If this is meant to force resetting the size to minsize, you can use `layer_rename->reset_size();` instead.

akien-mga commented on editor/editor_properties.h:
```diff
@@ -279,7 +279,44 @@ class EditorPropertyFlags : public EditorProperty {
+	void gui_input(const Ref<InputEvent> &p_ev) override;
+	void _notification(int p_what);
+	void set_flag(uint32_t p_flag);
+	static void _bind_methods();
```
`_notification` and `_bind_methods` should be `protected`, not `public`.

akien-mga commented on editor/editor_properties.h:
```diff
@@ -279,7 +279,44 @@ class EditorPropertyFlags : public EditorProperty {
+	Vector<String> names;
+	Vector<String> tooltips;
+
+	EditorPropertyLayersGrid();
```
Nitpick but usually we list the constructor at the end.

akien-mga commented on editor/editor_properties.h:
```diff
@@ -297,19 +334,22 @@ class EditorPropertyLayers : public EditorProperty {
 
 public:
 	void setup(LayerType p_layer_type);
+	void set_layer_name(const int p_index, const String name);
```
```suggestion
	void set_layer_name(int p_index, const String &p_name);
```
And same in `.cpp`.

akien-mga commented on editor/editor_properties.h:
```diff
@@ -279,7 +279,44 @@ class EditorPropertyFlags : public EditorProperty {
+	ConfirmationDialog *rename_dialog;
+	LineEdit *rename_dialog_text;
+
+	void _rename_pressed(const int p_menu);
```
`const` is overkill for `int` parameters (at least that's our current consensus :))
```suggestion
	void _rename_pressed(int p_menu);
```

akien-mga commented:
Thanks!
